File: 1_easy/00283_move_zeroes/two_pointers.cpp
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int len = nums.size();
        int k = 0;

        for (int i{0}; i < len; ++i) {
            if (nums[i] != 0) {
                nums[k] = nums[i];
                k += 1;
            }
        }

        for (int i{k}; i < len; ++i) {
            nums[i] = 0;
        }
    }
};

File: 2_medium/0001_1000/00189_rotate_array/two_pointers.rs
impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let len = nums.len();
        let k = k as usize % len;

        let mut left = 0_usize;
        let mut right = len - k - 1;

        while left < right {
            let temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left += 1;
            right -= 1;
        }

        left = len - k;
        right = len - 1;

        while left < right {
            let temp = nums[left];
            nums[left] = nums[right];
            nums[right] = temp;
            left += 1;
            right -= 1;
        }

        nums.reverse();
    }
}
